### ISSUE TITLE
Add some way to add custom wheel and dependencies to existing pyodide

### ISSUE THREAD

**Issue by joemarshall** [pyodide/pyodide#2857]
Importing externally built wheels is pretty easy, you can just call loadPackage on them. But because they're not in the packages.json, they don't get dependencies loaded correctly.

Micropip in theory would work, although it only works if the wheel and all dependencies are pure python. It also doesn't work if some dependencies are in packages which are manually loaded. And if there are some dependencies which you know can be skipped for your use case, but are listed on pypi, again you're stuck.

What I really want personally is just an API to add an entry to be the package list for a wheel URL plus dependencies. So I can say e.g. 
```
pyodide.addPackageEntry({name:"name",file:"http://whatever", deps:["some package"]}
```
It would make it easier to bring just new packages in from my own build, rather than having to host pyodide myself (or as a minimum to mirror and alter packages.json)

It seems to me like this would also be dead useful once it is possible to build pyodide packages without doing a full pyodide build.

Is this a stupid idea? Is there something I'm missing that means this isn't needed?

**Comment by hoodmane** [comment]
Pyodide 0.21.0 will come with fixes for a lot of this, but there is still more work to be done. The current status:
* Micropip can load binary wheels
* We added [`micropip.freeze`](https://pyodide.org/en/latest/usage/api/micropip-api.html#micropip.freeze) which creates a `repodata.json` (formerly `packages.json`) file
* you can specify a custom `repodata.json` file with the `lockFileURL` argument to `loadPyodide`

Particularly the custom lock file should make it not necessary to host your own copy of Pyodide. In addition, my hope is that my recent work on out of tree package builds should mean that even when creating binary extensions, you don't need to build Pyodide.

But we still don't have anything to add a new package at runtime to the lock file, nor any convenient way to access the lock file. I think something like `addPackageEntry` would be useful. I would personally like to add an `addPackageEntry` API and then remove from `loadPackage` the capability of loading from a URL in favor of `addPackageEntry({name : "name", file: url, deps: []})`.

> if there are some dependencies which you know can be skipped for your use case, but are listed on pypi, again you're stuck.

Now you can manually write a lockfile. I think if you use `micropip.install(pkg, deps=False)` then `micropip.freeze` will create a package entry and then you can delete the dependencies you don't want. But I haven't tested the interaction between `deps=False` and `micropip.freeze` it is possible that they interact badly.

**Comment by joemarshall** [comment]
Looking at the code addPackageEntry seems like a pretty trivial change, just needs to add to the two dicts (the main package list and the import name to package one)?

Good to know about the lockfileUrl option though, I'll probably put a pr into jupyterlite at some point to pass that through from the config file.

**Comment by hoodmane** [comment]
> addPackageEntry seems like a pretty trivial change

Yup. Presumably we should also validate that the deps are already existing packages. And what about dependency overrides?
